src/theory/strings/theory_strings_preprocess.cpp

```diff
@@ -363,28 +399,65 @@ Node StringsPreprocess::simplify( Node t, std::vector< Node > &new_nodes ) {
             NodeManager::currentNM()->mkNode(kind::OR, cc1, cc2), cc3);
     new_nodes.push_back( conc );
     retNode = pret;
-  } else if( t.getKind() == kind::STRING_STRREPL ) {
+  }
+  else if (t.getKind() == kind::STRING_STRREPL)
+  {
+    // processing term: replace( x, y, z )
     Node x = t[0];
     Node y = t[1];
     Node z = t[2];
-    Node sk1 = NodeManager::currentNM()->mkSkolem( "rp1", t[0].getType(), "created for replace" );
-    Node sk2 = NodeManager::currentNM()->mkSkolem( "rp2", t[0].getType(), "created for replace" );
-    Node skw = NodeManager::currentNM()->mkSkolem( "rpw", t[0].getType(), "created for replace" );
-    Node cond = NodeManager::currentNM()->mkNode( kind::STRING_STRCTN, x, y );
-    cond = NodeManager::currentNM()->mkNode( kind::AND, cond, NodeManager::currentNM()->mkNode(kind::GT, NodeManager::currentNM()->mkNode(kind::STRING_LENGTH, y), d_zero) );
-    Node c1 = x.eqNode( NodeManager::currentNM()->mkNode( kind::STRING_CONCAT, sk1, y, sk2 ) );
-    Node c2 = skw.eqNode( NodeManager::currentNM()->mkNode( kind::STRING_CONCAT, sk1, z, sk2 ) );
-    Node c3 = NodeManager::currentNM()->mkNode(kind::STRING_STRCTN,
-                NodeManager::currentNM()->mkNode(kind::STRING_CONCAT, sk1,
-                   NodeManager::currentNM()->mkNode(kind::STRING_SUBSTR, y, d_zero,
-                      NodeManager::currentNM()->mkNode(kind::MINUS,
-                        NodeManager::currentNM()->mkNode(kind::STRING_LENGTH, y),
-                        NodeManager::currentNM()->mkConst(::CVC4::Rational(1))))), y).negate();
-    Node rr = NodeManager::currentNM()->mkNode( kind::ITE, cond,
-                                                NodeManager::currentNM()->mkNode( kind::AND, c1, c2, c3),
-                                                skw.eqNode(x) );
+    TypeNode tn = t[0].getType();
+    Node rp1 = nm->mkSkolem("rp1", tn, "created for replace");
+    Node rp2 = nm->mkSkolem("rp2", tn, "created for replace");
+    Node rpw = nm->mkSkolem("rpw", tn, "created for replace");
+
```

Comment:
We could assert that len(rpw) <= len(x) + len(z), not sure how helpful that would be though.

Reply:
Right, this is analogous to the additional assertions for the indexof reduction.  Conversely, we could test removing those :) 

In the past I went through many iterations of playing with these, so I'm assuming it's currently close to optimal.  We could retest this, but there are higher priority items to deal with in my opinion.